### ISSUE TITLE
Define a Sigstore root of trust 

### ISSUE THREAD

**Issue by asraa** [sigstore/protobuf-specs#4]
**Define a Sigstore root of trust protobuf**

This issue is to track a protobuf definition of a Sigstore-specific "trust root". This would be used by clients as an input to verify an artifact/signature, as the only other input besides the artifact target information (which may be an offline OR online bundle of information). The information would contain:
* Rekor: A map of Rekor public keys, keyed by the log ID and containing extra custom metadata (e.g. online date range, identifiable URI, and more to be extended)
* Fulcio: A map of Fulcio certificates. I'm not sure what to key these by. We'd want the online date range, URI, subject/issuer. The online date range could be the most useful thing to use for identifying which Fulcio root CA was used to sign off on a particular signature (all Fulcio-issued certs are short-lived, so time is necessary in verification). TODO: How to represent intermediates? Intermediates *should* be alongside the roots, and there may be a collection. A "Fulcio" target could be represented as a map[string][]*x509.Certificate (i.e. a list of certs)
* CTFE: A map of CT public keys, keyed by the log ID and containing the same extra custom metadata as Rekor. 
* And more to come (TSA)

The data structure for maps/list would imply that clients may need to hold on to multiple keys and certificates, because they may be validating targets signed on previous shards in past years. During verification, clients must use information in this struct to identify the relevant data for the particular target that it's verifying at runtime. 

In other words, this information is all **pre-resolved** per-instance of the client, which makes it use-able for a variety of different environments and types of trust roots (online TUF, hard-coded test variables, etc). The verification function signature would look something like `verify(bundle, trust_root)`

I propose calling it (please suggest!):
* `TrustedRoot`
* `SigstoreTrustRoot`
* `TrustedRootStore`

I think prefacing it with `Sigstore` might not be necessary; it's already defined in the Sigstore package. 

I'll give a quick impl start, but please comment here for any discussions/ideas that I should take into consideration before review.

@vaikas @kommendorkapten @joshuagl @patflynn @jku @haydentherapper @znewman01

**Comment by kommendorkapten** [comment]
Nice that you called this out @asraa, I'm actually planning to get a draft on this (in protobuf) this week. I there is no objection I can assign myself to it.

**Comment by asraa** [comment]
Sure!

**Comment by kommendorkapten** [comment]
As part of the bundle work I started to sketch on this: https://github.com/kommendorkapten/cosign/blob/bundle_verification/specs/dotsigstore_bundle/verify.md#input

I'm happy to collaborate on this if you want to! Or I just compile the data in this issue with the linked one and we take it from there, should be fairly easy to get a v1 PR that we can start to discuss around.

**Comment by asraa** [comment]
> Or I just compile the data in this issue with the linked one and we take it from there, should be fairly easy to get a v1 PR that we can start to discuss around.

Up to you! I have suggestions, but it's easiest for me to make suggestions on a PR or through docs (and might as well not disturb normal flow). Let's do a v1 PR and I can make suggestions commits.

**Comment by kommendorkapten** [comment]
I'll promise you a draft PR tomorrow before you wake up 💯

**Comment by haydentherapper** [comment]
> Intermediates should be alongside the roots 

+1 

> CTFE: A map of CT public keys, keyed by the log ID and containing the same extra custom metadata as Rekor.

We should also make sure we can change how we represent the log ID in the future (a couple of years out I bet). RFC9162 changed the log ID format from a hash to an OID.

**Comment by asraa** [comment]
> We should also make sure we can change how we represent the log ID in the future (a couple of years out I bet). RFC9162 changed the log ID format from a hash to an OID.

Agree, this would be agnostic of the calculation type, as long as it matches the LogId field in the transparency log entry. Definitely we should not require that the log id is hash(pubkey)

**Comment by jku** [comment]
This is maybe only tangentially related but I would also like a clear list of steps that a client should follow to initialize and to use their trust root -- and I am not talking about mechanisms like protos or TUF but human readable instructions for what a client should do with this bundle:
* what is the expected input that starts the root trust cache update (how does client know which sigstore instance to use, do clients even need to support multiple instances)
* recommendations for cases where "multiple trust roots" are in use: e.g. how does client decide which fulcio certificate to use when signing? Should they use all fulcio certificates in the bundle for verifying signatures?
* does the trust root define the various service endpoints the "sigstore instance" provides or are they somehow known beforehand?

**Comment by kommendorkapten** [comment]
@jku Yes, that is a very important, and not something we have documented today. I have started to draft details on this in the [client spec](https://docs.google.com/document/d/1kbhK2qyPPk8SLavHzYSDM8-Ueul9_oxIMVFuWMWKz0E/edit#heading=h.x8bduppe89ln) but it's still very much WIP.

**Comment by jku** [comment]
Oh sure, it's just that as a relative sigstore outsider I have trouble confirming whether the proposed proto contains too little, too much or just the right amount of information because the correct usage is not codified and I can't really trust the current client implementations -- it's part of the reason I've not contributed more yet.
